universal-application-tool-0.0.1/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -23,28 +24,81 @@
 import services.question.ScalarType;
 import services.question.UnsupportedScalarTypeException;
 
 public class ApplicantServiceImpl implements ApplicantService {
 
   private final ApplicantRepository applicantRepository;
   private final ProgramService programService;
+  private final Clock clock;
   private final HttpExecutionContext httpExecutionContext;
+  private static final ImmutableSet<String> RESERVED_SCALAR_KEYS =
+      ImmutableSet.of(
+          QuestionDefinition.METADATA_UPDATE_TIME_KEY,
+          QuestionDefinition.METADATA_UPDATE_PROGRAM_ID_KEY);
   private final Logger log = LoggerFactory.getLogger(ApplicantService.class);
 
   @Inject
   public ApplicantServiceImpl(
       ApplicantRepository applicantRepository,
       ProgramService programService,
+      Clock clock,
       HttpExecutionContext httpExecutionContext) {
     this.applicantRepository = checkNotNull(applicantRepository);
     this.programService = checkNotNull(programService);
+    this.clock = checkNotNull(clock);
     this.httpExecutionContext = checkNotNull(httpExecutionContext);
   }
 
+  @Override
+  public CompletionStage<Applicant> createApplicant(long userId) {
+    Applicant applicant = new Applicant();
+    return applicantRepository.insertApplicant(applicant).thenApply((unused) -> applicant);
+  }
+
+  @Override
+  public CompletionStage<ReadOnlyApplicantProgramService> getReadOnlyApplicantProgramService(
+      long applicantId, long programId) {
+    CompletableFuture<Optional<Applicant>> applicantCompletableFuture =
+        applicantRepository.lookupApplicant(applicantId).toCompletableFuture();
+    CompletableFuture<ProgramDefinition> programDefinitionCompletableFuture =
+        programService.getProgramDefinitionAsync(programId).toCompletableFuture();
+
+    return CompletableFuture.allOf(applicantCompletableFuture, programDefinitionCompletableFuture)
+        .thenApplyAsync(
+            (v) -> {
+              Applicant applicant = applicantCompletableFuture.join().get();
+              ProgramDefinition programDefinition = programDefinitionCompletableFuture.join();
+
+              return new ReadOnlyApplicantProgramServiceImpl(
+                  applicant.getApplicantData(), programDefinition);
+            },
+            httpExecutionContext.current());
+  }
+
   @Override
   public CompletionStage<ErrorAnd<ReadOnlyApplicantProgramService, Exception>>
+      stageAndUpdateIfValid(
+          long applicantId, long programId, long blockId, ImmutableMap<String, String> updateMap) {
+    ImmutableSet<Update> updates =
+        updateMap.entrySet().stream()
+            .map(entry -> Update.create(Path.create(entry.getKey()), entry.getValue()))
+            .collect(ImmutableSet.toImmutableSet());
+
+    boolean updatePathsContainReservedKeys =
+        updates.stream().anyMatch(u -> RESERVED_SCALAR_KEYS.contains(u.path().keyName()));
+    if (updatePathsContainReservedKeys) {
+      return CompletableFuture.completedFuture(
```

Comment:
Suggestion (blocking): If this is not an expected user input error, please raise an exception instead.

```
return CompletableFuture.failedFuture(new IllegalArgumentException("Path contained reserved scalar key"));
```

Please handle the exception in the controller.

```
.exceptionally(
  ex -> {
    if (ex instanceof CompletionException) {
                Throwable cause = ex.getCause();
                if (cause ...) {
                   ...
                } else if (cause instanceof IllegalArgumentException) {
                  // handle this exception
                  return badRequest("Unable to process this request");
                }
                // ...
     }
  });
```

Reply:
Thanks Hung-Ming! I updated to throw instead of return ErrorAnd, mind double-checking my work?

Reply:
Looks great! Thanks for doing this!